pkg/agent/controllers/agent/agent.go

```diff
@@ -0,0 +1,205 @@
+
+		excludeNamespaces: exclusionNSMap,
+	}
+
+	klog.Info("Setting up event handlers")
+	config.ServiceInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc: agentController.enqueueService,
+		UpdateFunc: func(old, newObj interface{}) {
+			agentController.handleUpdateService(old, newObj)
+		},
+		DeleteFunc: agentController.handleRemovedService,
+	}, handlerResync)
+
+	return agentController, nil
+}
+
+func (a *Controller) Run(stopCh <-chan struct{}) error {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer utilruntime.HandleCrash()
+
+	// Start the informer factories to begin populating the informer caches
+	klog.Info("Starting IPAM Controller")
+
+	// Wait for the caches to be synced before starting workers
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, a.servicesSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	go wait.Until(a.runServiceWorker, time.Second, stopCh)
+	wg.Wait()
+	klog.Info("Lighthouse agent workers started")
+	<-stopCh
+	klog.Info("Shutting down workers")
+	return nil
+}
+
+func (a *Controller) enqueueService(obj interface{}) {
+	if key := a.getEnqueueKey(obj); key != "" {
+		klog.V(4).Infof("Enqueueing service %v for agent", key)
+		a.serviceWorkqueue.AddRateLimited(key)
+	}
+}
+
+func (a *Controller) getEnqueueKey(obj interface{}) string {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return ""
+	}
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return ""
+	}
+	if a.excludeNamespaces[ns] || excludeSvcList[name] {
+		return ""
+	}
+	return key
+}
+func (a *Controller) runServiceWorker() {
+	for a.processNextService() {
+
+	}
+}
+
+func (a *Controller) processNextService() bool {
+	obj, shutdown := a.serviceWorkqueue.Get()
+	if shutdown {
+		return false
+	}
+	err := func() error {
+		defer a.serviceWorkqueue.Done(obj)
+
+		key := obj.(string)
+		ns, name, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			a.serviceWorkqueue.Forget(obj)
+			return fmt.Errorf("error while splitting meta namespace key %s: %v", key, err)
+		}
+
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			// Retrieve the latest version of object before attempting update
+			// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+			service, err := a.kubeClientSet.CoreV1().Services(ns).Get(name, metav1.GetOptions{})
+
+			if err != nil {
+				if errors.IsNotFound(err) {
+					// already deleted Forget and return
+					a.serviceWorkqueue.Forget(obj)
+				} else {
+					// could be API error, so we want to requeue
+					logAndRequeue(key, a.serviceWorkqueue)
+				}
+				return fmt.Errorf("error retrieving service %s: %v", name, err)
+			}
+
+			return a.serviceUpdater(service, key)
+		})
+		if retryErr != nil {
+			return retryErr
```

Comment:
If max retries was reached then we'd want to either call `Forget` or re-queue. I'd suggest to return a nil error above for the `errors.IsNotFound(err)` case, then here re-queue for any error returned.

Reply:
For `errors.IsNotFound` we want to Forget coz it is expected to get that error when object is already deleted. If we keep requeuing on max retries, it kinda defeats point of a max retries. I think we can revisit this if we run into issues during scale/perf testing. Thoughts?

Reply:
Either way you would still `Forget` on L126. But you also need to `Forget` if max retries is reached - it's really a bug otherwise. But I'm suggesting to simplify the code and just re-queue on any error here, which means `IsNotFound` above would return nil error and also not log since it's not really an error.  If max retries is reached I think we still want to retry later (see comment in `handleUpdateService`).

Reply:
Aha, got it. Will do.